src/dbnode/storage/series/buffer.go

```diff
@@ -473,119 +460,326 @@ func (b *dbBuffer) FetchBlocksMetadata(
-	encoders          []inOrderEncoder
-	bootstrapped      []block.DatabaseBlock
-	lastReadUnixNanos int64
-	drained           bool
+func (b *dbBuffer) bucketVersionsAt(
+	t time.Time,
+) (*BufferBucketVersions, bool) {
+	// First check LRU cache.
+	for _, buckets := range b.bucketVersionsCache {
+		if buckets == nil {
+			continue
+		}
+		if buckets.start.Equal(t) {
+			return buckets, true
+		}
+	}
+
+	// Then check the map.
+	if buckets, exists := b.bucketsMap[xtime.ToUnixNano(t)]; exists {
+		return buckets, true
+	}
+
+	return nil, false
 }
 
-type inOrderEncoder struct {
-	encoder     encoding.Encoder
-	lastWriteAt time.Time
+func (b *dbBuffer) bucketsAtCreate(
+	t time.Time,
+) *BufferBucketVersions {
+	if buckets, exists := b.bucketVersionsAt(t); exists {
+		return buckets
+	}
+
+	buckets := b.bucketVersionsPool.Get()
+	buckets.resetTo(t, b.opts, b.bucketPool)
+	b.bucketsMap[xtime.ToUnixNano(t)] = buckets
+	b.inOrderBlockStartsAdd(t)
+
+	return buckets
 }
 
-func (b *dbBufferBucket) resetTo(
-	start time.Time,
-) {
-	// Close the old context if we're resetting for use
-	b.finalize()
+func (b *dbBuffer) putBucketInCache(newBuckets *BufferBucketVersions) {
+	replaceIdx := bucketsCacheSize - 1
+	for i, buckets := range b.bucketVersionsCache {
+		// Check if we have the same pointer in cache.
+		if buckets == newBuckets {
+			replaceIdx = i
+		}
+	}
 
-	bopts := b.opts.DatabaseBlockOptions()
-	encoder := bopts.EncoderPool().Get()
-	encoder.Reset(start, bopts.DatabaseBlockAllocSize())
+	for i := replaceIdx; i > 0; i-- {
+		b.bucketVersionsCache[i] = b.bucketVersionsCache[i-1]
+	}
+
+	b.bucketVersionsCache[0] = newBuckets
+}
+
+func (b *dbBuffer) removeBucketsAt(blockStart time.Time) {
+	buckets, exists := b.bucketVersionsAt(blockStart)
+	if !exists {
+		return
+	}
+	// nil out pointers.
+	buckets.resetTo(timeZero, nil, nil)
+	b.bucketVersionsPool.Put(buckets)
+	delete(b.bucketsMap, xtime.ToUnixNano(blockStart))
+	b.inOrderBlockStartsRemove(blockStart)
+}
+
+func (b *dbBuffer) inOrderBlockStartsAdd(newTime time.Time) {
+	starts := b.inOrderBlockStarts
+	idx := len(starts)
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.After(newTime) {
+			idx = i
+			break
+		}
+	}
+	// Insert new time without allocating new slice.
+	b.inOrderBlockStarts = append(starts, timeZero)
+	// Update to new slice
+	starts = b.inOrderBlockStarts
+	copy(starts[idx+1:], starts[idx:])
+	starts[idx] = newTime
+}
+
+func (b *dbBuffer) inOrderBlockStartsRemove(removeTime time.Time) {
+	starts := b.inOrderBlockStarts
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.Equal(removeTime) {
+			b.inOrderBlockStarts = append(starts[:i], starts[i+1:]...)
+			return
+		}
+	}
+}
 
+// BufferBucketVersions is a container for different versions (from
+// different flushes to disk) of buffer buckets in the database.
+type BufferBucketVersions struct {
```

Comment:
Does this need to be an exported type?

Reply:
Since we want to have a database-level pool of these, we use this when passing pools via our Options.

Reply:
I see, makes sense.